File: archive/test_items_backup/audit_extras.py
"""Build the contact sheets for audit_extras.mjs.

Each unaccounted-for label becomes one numbered tile: the sticker cropped out of
the NATIVE photo with enough margin to show its neighbours, captioned with what
the app read. Looking at the tile answers the only question that matters — is
there really a sticker there saying that, or did the app invent a book?

    node audit_extras.mjs        # -> _audit/_plan.json
    python audit_extras.py       # -> _audit/<img>.jpg

The bbox in the plan is in the coordinate space of the image the app UPLOADED,
which encodeForOCR caps at maxDim 2400. Every test photo is 4080 on its long
edge, so the scale back to native is 4080/2400 = 1.70 — but it is computed per
image rather than assumed, because assuming it is exactly the bug that produced
two useless crops last time (the portrait photo 195849222 is 3072x4080, so
width/2400 is wrong for it).
"""
import json, os, sys
from PIL import Image, ImageDraw

HERE = os.path.dirname(os.path.abspath(__file__))
OUT = os.path.join(HERE, '_audit')
PLAN = '_plan.json'
SUFFIX = ''
for a in sys.argv[1:]:
    if a.startswith('--plan='):
        PLAN = a.split('=', 1)[1]
        SUFFIX = '_' + os.path.splitext(PLAN)[0].strip('_')
SENT_MAX_DIM = 2400          # encodeForOCR's cap
PAD_X, PAD_ABOVE, PAD_BELOW = 2.6, 2.2, 3.2   # in sticker-heights
COLS = 6
TILE_W, TILE_H = 300, 260
CAPTION_H = 34

plan = json.load(open(os.path.join(OUT, PLAN)))

for img_name, rec in plan.items():
    extras = rec['extras']
    if not extras:
        print(f'{img_name}: nothing to audit')
        continue
    im = Image.open(os.path.join(HERE, img_name))
    # maxDim, NOT width — the portrait photo would scale wrong otherwise.
    scale = max(im.width, im.height) / SENT_MAX_DIM

    tiles = []
    for e in extras:
        x, y, w, h = e['bbox']
        x0 = int((x - h * PAD_X) * scale)
        y0 = int((y - h * PAD_ABOVE) * scale)
        x1 = int((x + w + h * PAD_X) * scale)
        y1 = int((y + h * (1 + PAD_BELOW)) * scale)
        # Where the sticker itself sits inside the crop, so the tile can mark it.
        box = (int(x * scale) - x0, int(y * scale) - y0,
               int((x + w) * scale) - x0, int((y + h) * scale) - y0)
        x0, y0 = max(0, x0), max(0, y0)
        x1, y1 = min(im.width, x1), min(im.height, y1)
        if x1 - x0 < 8 or y1 - y0 < 8:
            tiles.append((None, e))
            continue
        crop = im.crop((x0, y0, x1, y1))
        d = ImageDraw.Draw(crop)
        d.rectangle(box, outline=(255, 40, 40), width=3)
        crop.thumbnail((TILE_W, TILE_H - CAPTION_H), Image.LANCZOS)
        tiles.append((crop, e))

    rows = (len(tiles) + COLS - 1) // COLS
    sheet = Image.new('RGB', (COLS * TILE_W, rows * TILE_H), 'white')
    dr = ImageDraw.Draw(sheet)
    for i, (crop, e) in enumerate(tiles):
        cx, cy = (i % COLS) * TILE_W, (i // COLS) * TILE_H
        if crop:
            sheet.paste(crop, (cx + (TILE_W - crop.width) // 2, cy + CAPTION_H))
        cap = f"{i + 1}. {e['text']}"
        if e.get('score') is not None:
            cap += f"  ({e['score']:.2f})"
        dr.text((cx + 6, cy + 6), cap, fill=(0, 0, 0))
        dr.rectangle([cx, cy, cx + TILE_W - 1, cy + TILE_H - 1], outline=(200, 200, 200))

    dst = os.path.join(OUT, img_name.replace('.jpg', '') + SUFFIX + '.jpg')
    sheet.save(dst, quality=88)
    print(f'{img_name}: {len(tiles)} crop(s) -> {dst}  ({sheet.width}x{sheet.height})')

File: archive/test_items_backup/audit_extras.mjs
/**
 * Crop every label the app emitted that ground truth does NOT account for.
 *
 * Why this exists: the app emits ~388 labels across the 10 shelves and only
 * ~284 match an entry in ground_truth_v2.json. That gap is NOT 104 mistakes —
 * some of it is real books the answer key never recorded (635 GAR, 635 CAL,
 * 635 BAK and 635.0484 BEN were all found that way), and some is genuine junk
 * (spine titles and frame-edge fragments read as call numbers: "635 ICON",
 * "604 FOTO", "635 AK"). Precision cannot be quoted until the two are told
 * apart, and the only way to tell them apart is to look at the pixels.
 *
 * So: cut each unaccounted-for label out of the native photo, and lay them on
 * one numbered contact sheet per image. A human (or a vision model) reads the
 * sheet and says, for each crop, whether a real sticker is there.
 *
 *   node audit_extras.mjs --fixture=live.json      # -> _audit/<img>.jpg + .json
 *
 * The matcher is COPIED from site_suite.mjs on purpose: this tool has to ask
 * "what did the suite fail to match", so it must use the suite's own rule. If
 * that rule changes, change it here too — they are meant to agree.
 */
import { readFileSync, writeFileSync, mkdirSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { execFileSync } from 'child_process';
import { loadGT } from './gt_load.mjs';

const HERE = dirname(fileURLToPath(import.meta.url));
const arg = (k, d = '') =>
  (process.argv.find(a => a.startsWith(`--${k}=`)) || '').split('=').slice(1).join('=') || d;

const FIXTURE = arg('fixture', 'live.json');
const OUT = join(HERE, '_audit');
mkdirSync(OUT, { recursive: true });

// ── site_suite's matcher, verbatim ──────────────────────────────────────────
function parseDeweyStr(s) {
  const m = /^\s*(\d+(?:\.\d+)?)\s*([A-Za-z]{1,4})?/.exec(String(s));
  return m ? { num: parseFloat(m[1]), cut: (m[2] || '').toUpperCase() } : null;
}
function numStr(s) { const m = /^\s*(\d+(?:\.\d+)?)/.exec(String(s)); return m ? m[1] : null; }
function surname(s) {
  s = String(s).trim().toUpperCase();
  const m = /^(VON\s+\w+|VAN\s+\w+|DE\s+\w+|LA\s+\w+)/.exec(s);
  if (m) return m[1].replace(/\s+/g, '');
  return s.split(/[,\s]+/)[0] || '';
}
function labelMatch(detected, gt, section) {
  if (section === 'nonfiction') {
    const dn = numStr(detected), gn = numStr(gt);
    const dm = parseDeweyStr(detected), gm = parseDeweyStr(gt);
    if (dn && gn && dm && gm) {
      const numOk = dn.startsWith(gn) || gn.startsWith(dn);
      const dc = (dm.cut || '').slice(0, 3), gc = (gm.cut || '').slice(0, 3);
      return numOk && !!dc && !!gc && dc === gc;
    }
    return false;
  }
  const ds = surname(detected), gs = surname(gt);
  const mn = Math.min(ds.length, gs.length);
  if (mn < 3) return ds === gs;
  if (mn >= 6) { let d = 0; for (let i = 0; i < mn; i++) if (ds[i] !== gs[i]) d++; return d <= 1; }
  return ds.slice(0, 4) === gs.slice(0, 4);
}

// ── Which emitted labels does GT not account for? ───────────────────────────
const GT = loadGT();
const live = JSON.parse(readFileSync(join(HERE, FIXTURE), 'utf-8'));

const plan = {};
let totalExtra = 0, totalEmitted = 0;
for (const r of (live.results || [])) {
  const g = GT[r.imgName];
  if (!g || !r.fixture) continue;
  const gtLabels = [];
  for (const row of (g.rows || [])) gtLabels.push(...row.labels);

  // Greedy one-to-one, same as matchToGt: a GT entry can only absorb one read.
  const used = new Set();
  gtLabels.forEach(gl => {
    for (let di = 0; di < r.fixture.length; di++) {
      if (used.has(di)) continue;
      if (labelMatch(r.fixture[di].spine_label, gl, r.section)) { used.add(di); break; }
    }
  });

  const extras = r.fixture
    .map((b, i) => ({ ...b, i }))
    .filter(b => !used.has(b.i) && b.spine_label && b._bbox);
  totalExtra += extras.length;
  totalEmitted += r.fixture.length;
  plan[r.imgName] = {
    section: r.section,
    emitted: r.fixture.length,
    accountedFor: used.size,
    extras: extras.map(b => ({
      text: b.spine_label, score: b._score ?? null, src: b._src ?? null,
      title: b.title ?? null, bbox: b._bbox,
    })),
  };
  console.log(`${r.imgName.replace('PXL_2026060', '').padEnd(20)} ` +
    `emitted ${String(r.fixture.length).padStart(3)}  ` +
    `matched ${String(used.size).padStart(3)}  ` +
    `unaccounted ${String(extras.length).padStart(3)}`);
}
console.log(`\n${totalExtra} of ${totalEmitted} emitted labels are not in ground truth.`);

writeFileSync(join(OUT, '_plan.json'), JSON.stringify(plan, null, 1));
console.log(`Wrote ${join(OUT, '_plan.json')}`);
console.log('Now run:  python audit_extras.py     # -> _audit/<img>.jpg contact sheets');
